language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -220,47 +242,86 @@ typeOf (TyInst x body ty) = do
         _ -> throwError (TypeMismatch x (void body) (TyForall () dummyTyName dummyKind dummyType) nBodyTy)
 typeOf (Unwrap x body) = do
     nBodyTy@(NormalizedType bodyTy) <- typeOf body
     case bodyTy of
         TyFix _ n fixTy -> do
-            let subst = tySubstitute (extractUnique n) nBodyTy (NormalizedType fixTy)
-            pure (tyReduce subst)
+            tyEnvAssign (extractUnique n) nBodyTy
+            tyReduce fixTy <* tyEnvDelete (extractUnique n)
         _             -> throwError (TypeMismatch x (void body) (TyFix () dummyTyName dummyType) nBodyTy)
 typeOf (Wrap x n ty body) = do
     nBodyTy <- typeOf body
-    let fixed = tySubstitute (extractUnique n) (NormalizedType $ TyFix () (void n) (void ty)) (void $ NormalizedType ty)
+    tyEnvAssign (extractUnique n) (NormalizedType $ TyFix () (void n) (void ty))
     typeCheckStep
-    if tyReduce fixed == nBodyTy
-        then pure $ NormalizedType (TyFix () (void n) (void ty)) -- type annotations on terms must be normalized
-        else throwError (TypeMismatch x (void body) fixed nBodyTy)
+    red <- tyReduce (void ty) <* tyEnvDelete (extractUnique n)
+    if red == nBodyTy
+        then pure $ NormalizedType (TyFix () (void n) (void ty))
+        else throwError (TypeMismatch x (void body) (getNormalizedType red) nBodyTy)
 
 extractUnique :: TyNameWithKind a -> Unique
 extractUnique = nameUnique . unTyName . unTyNameWithKind
 
--- TODO: make type substitutions occur in a state monad + benchmark
--- | Substitute the given type into another type. The input types must be normalized - this prevents us creating redundant work. The
--- output type is not normalized, since we may introduce new redexes.
-tySubstitute :: Unique -- ^ Unique associated with type variable
-             -> NormalizedType TyNameWithKind a -- ^ Type we are binding to free variable
-             -> NormalizedType TyNameWithKind a -- ^ Type we are substituting in
-             -> Type TyNameWithKind a
-tySubstitute u (NormalizedType ty) = cata a . getNormalizedType where
-    a (TyVarF _ (TyNameWithKind (TyName (Name _ _ u')))) | u == u' = ty
-    a x                                                  = embed x
-
--- also this should involve contexts
+tyEnvDelete :: MonadState (TypeSt, Natural) m
+            => Unique
+            -> m ()
+tyEnvDelete (Unique i) = modify (first (IM.delete i))
+
+tyEnvAssign :: MonadState (TypeSt, Natural) m
+            => Unique
+            -> NormalizedType TyNameWithKind ()
+            -> m ()
+tyEnvAssign (Unique i) ty = modify (first (IM.insert i ty))
+
+isTyLam :: Type TyNameWithKind () -> Bool
+isTyLam TyLam{} = True
+isTyLam _       = False
+
+maybeRed :: Bool -> Type TyNameWithKind () -> TypeCheckM a (NormalizedType TyNameWithKind ())
+maybeRed True  = tyReduce
+maybeRed False = pure . NormalizedType
+
+-- This function is necessary because type reductions do *not* traverse 'TyFix'
+-- but context rewrites must traverse *all* constructors.
```

Comment:
Please describe why type reductions do not traverse `TyFix` and why it makes sense here. What does this function do?

Reply:
This follows the spec.

Reply:
Is this deliberate in the spec?

I'd prefer if we had a self-contained type checking algorithm rather than having implicit (or explicit) references to the spec. If there is a reason not to traverse `TyFix`, I think we should include it here.